edb/common/ast/base.py

```diff
@@ -219,24 +232,19 @@ def __init__(self, **kwargs):
                 f'cannot instantiate abstract AST node '
                 f'{self.__class__.__name__!r}')
 
-        should_check_types = __debug__ and _check_type is _check_type_real
-        for field_name, field in self.__class__._fields.items():
-            if field_name in kwargs:
-                value = kwargs[field_name]
-            elif field.factory is not None:
-                value = field.factory()
-            else:
-                value = field.default
+        # Put things directly into __dict__, as an optimization
+        d = self.__dict__
 
-            if should_check_types:
-                self.check_field_type(field, value)
+        for field_name, factory in self._field_factories:
+            value = factory()
```

Comment:
I'm not sure if unconditionally calling `factory()` is better than checking if `field_name not in kwargs` first. I mean it can be slightly faster until someone adds a more expensive factory.

Reply:
Can you add an if check and see if it costs too much? If it does we might as well compile this file with Cython.

Reply:
It's a tiny bit worse but it's pretty marginal

Reply:
Oh, but if we do that, then we can actually just make the updates directly to `kwargs`, then stick `kwargs` into `__dict__`, which gets us a bit more of a speedup

Reply:
Even better